Issue #30 - add lint dirty files only flag

sergesemashko opened the pull request:
- add a flag requested in the https://github.com/vieron/stylelint-webpack-plugin/issues/30
- alpha-sorted options in readme

@JaKXz, please, take a look

cc/ @athyuttamre

JaKXz commented:
@sergesemashko thanks for taking a crack at this. My first question is from the coveralls failure: can you think of ways to test this? 

Otherwise I like the way this is going.

JaKXz commented on index.js:
```diff
@@ -28,10 +28,36 @@ function apply(options, compiler) {
+
+      var changedFiles = Object.keys(compilation.fileTimestamps).filter(function (watchfile) {
+        return (this.prevTimestamps[watchfile] || this.startTime) < (compilation.fileTimestamps[watchfile] || Infinity);
+      }.bind(this)).filter(minimatch.filter(globPatterb.join('|'), {matchBase: true}));
```
Can this be re-writtern with a `reduce` on `Object.keys(compilation.fileTimestamps)`? it would be a bit clearer IMO and would be fewer calls/binds.

JaKXz replied:
You could even bring in `lodash.reduce` and split out this function into a private module in `lib`, since it's quite complex, and probably will have some compatibility issues with webpack 2 (#46).

sergesemashko replied:
Sorry, did not quite get where to put reduce. Would this be some like this?
```js
function getChangedFiles(plugin, compilation, glob) {
    var updatedStyles = reduce(compilation.fileTimestamps, function (changedStyleFiles, timestamp, filename) {
      if (
        // Check if file has been changed first ...
        (plugin.prevTimestamps[filename] || plugin.startTime) < (compilation.fileTimestamps[filename] || Infinity) &&
        // ... then validate by the glob pattern.
        minimatch(glob, {matchBase: true})
      ) {
        changedStyleFiles.push(filename);
      }
      return changedStyleFiles;
    }, []);
      
    return updatedStyles;
  }
```

An example without reduce:
```js
function getChangedFiles(plugin, compilation, glob) {
    // Filter only changed files first ...
    var updated = Object.keys(compilation.fileTimestamps).filter(function (filename) {
      return (plugin.prevTimestamps[filename] || plugin.startTime) < (compilation.fileTimestamps[filename] || Infinity);
    });
    // ... then filter files by the glob pattern.
    var updatedStyles = updated.filter(minimatch.filter(glob, {matchBase: true}));
    return updatedStyles;
  }
```

The actual call would look like:
```js
function apply(options, compiler) {
  var plugin = this;
  ...
  compiler.plugin('emit', function (compilation, callback) {
     ...
     var changedFiles = getChangedFiles(plugin, compilation, globPatterb.join('|'))
     ...
  }
  ...
}
```

@JaKXz, Let me what option you'd prefer to go with

JaKXz replied:
@sergesemashko I prefer the slightly purer first version, but I would use `.concat` instead of `.push`. I would also just return the result of `reduce(obj, fn, [])`.

JaKXz replied:
Also, the `emit` function should be bound to `this` so that it's consistent with the other lib functions.

sergesemashko replied:
in reduce iterates through the initial keys of `compilation.fileTimestamps`, so `filename` is there for sure

JaKXz replied:
👍 let's see how the implementation goes and discuss further there :)

JaKXz commented on package.json:
```diff
@@ -32,6 +32,7 @@
   "dependencies": {
     "arrify": "^1.0.1",
     "chalk": "^1.1.3",
+    "minimatch": "^3.0.3",
```
@sergesemashko just curious, is the filter fn that you're using from minimatch available from `node-glob`? could remove unnecessary dependencies :)

sergesemashko replied:
good catch. Minimatch is actually a dependency of `node-glob` which doesn't export any kind of function to use here :( But I ran `yarn add minimatch`, surprisingly, I did not cause any updates of `yarn.lock`, because latest minimatch version is already installed at `./node_modules`.

JaKXz replied:
Great!

sergesemashko commented on test/index.js:
```diff
@@ -174,4 +176,81 @@ describe('stylelint-webpack-plugin', function () {
+        } else
+        if (runsCount === 2) {
+          // Check that on JS file change styleling is not triggered.
+          expect(stats.compilation.errors).to.have.length(0);
```
@JaKXz, initially, I had separate test for testing for non-style file change. But seems like it's not possible to run multiple webpack watch instances simultaneously as tests are executed in parallel mode. So, I had to squash all watch cases into one.

JaKXz replied:
Can we just mock out the complexity and assert that we are seeing the right thing? I don't want to test webpack (the framework). 

It's definitely going to be easier to write a unit test for `get-changed-files`. I would start there in a new test for, and we can come back to the integration test later.

JaKXz commented on index.js:
```diff
@@ -28,10 +28,31 @@ function apply(options, compiler) {
+
+    compiler.plugin('emit', function (compilation, callback) {
+      var dirtyOptions = assign({}, options);
+      var globPatterb = dirtyOptions.files;
```
Nit: typo

JaKXz commented on index.js:
```diff
@@ -28,10 +28,31 @@ function apply(options, compiler) {
+    compiler.plugin('emit', function (compilation, callback) {
+      var dirtyOptions = assign({}, options);
+      var globPatterb = dirtyOptions.files;
+      var changedFiles = getChangedFiles(this, compilation, globPatterb.join('|'));
```
Can you pass an object here, naming the parameters? At some point I'll rewrite this plugin in the latest language features and use destructuring. For now it'll make it clearer why these args are being passed in. 

👍🏽 thanks for taking my comments well. Liking the way this is written :)

sergesemashko replied:
good call

JaKXz commented on test/helpers/watch.js:
```diff
@@ -0,0 +1,19 @@
+'use strict';
+
+var webpack = require('webpack');
```
Please rebase to capture our changes to support webpack 1&2.

sergesemashko commented:
@JaKXz, again, thanks for the feedback.
Didn't want to get rid of integration test completely (spent quite a lot of time on it:)), so I commented it out for later use with supporting comments.
Please, take a look once again.

JaKXz commented on test/index.js:
```diff
@@ -170,4 +171,117 @@ describe('stylelint-webpack-plugin', function () {
+    */
+  });
+
+  context('getChangedFiles', function () {
```
This should be in its own file. 👍 looks great though 😄

JaKXz commented on test/index.js:
```diff
@@ -170,4 +171,117 @@ describe('stylelint-webpack-plugin', function () {
+          - `../helpers/watch`
+          - `path`
+     // Webpack integration test.
+     it('lints only changed files in watch mode', function (done) {
```
This could be marked with `it.skip` so we don't have to commit commented out code. In fact it could be inside it's own `context` block.

In that block we should _stub_ out the parts of webpack that we can assume will operate as a black box.

Then, I believe the `watchCallback` can be [spied on and assertions can be made on individual calls of that fn](http://devdocs.io/sinon/index#sinon-spy). I don't mean to prescribe `sinon` as a solution since other libraries exist (e.g. I prefer [testdouble](https://github.com/testdouble/testdouble.js/)) but I just wanted to demonstrate that I think it's possible to write this test without all the imperative code and then we can merge it :)

sergesemashko replied:
good call, I'll try to mock blackbox logic then

JaKXz commented on test/fixtures/lint-dirty-files/updated-bad.scss:
```diff
@@ -0,0 +1,4 @@
+     #stuff
+{display: block
+;;
+   }
```
another nitpick: please make sure your editor is using [EditorConfig](http://editorconfig.org) - thanks!

JaKXz commented on index.js:
```diff
@@ -8,11 +8,11 @@ var formatter = require('stylelint').formatters.string;
 
 // Modules
 var runCompilation = require('./lib/run-compilation');
+var getChangedFiles = require('./lib/get-chaged-files');
```
nit: typo

JaKXz commented on lib/get-chaged-files.js:
```diff
@@ -0,0 +1,24 @@
+ * against cached timestamps from previous run.
+ *
+ * @param plugin - stylelint-webpack-plugin this scopr
+ * @param compilation
```
Do we need the entire `compilation` object here? If I haven't misread the code looks like we're just using the `fileTimestamps` property so I would just accept that as a param.

sergesemashko replied:
you are right, fileTimestamps should be enough

JaKXz commented:
@sergesemashko one thing I would suggest in general is to try to extract as many operations into pure "components" or functions that can be easily tested (like `get-changed-files`). Then maintenance (e.g. my life) becomes much easier because we just have to take care of small tested units rather than maintain tests of the integration of this complex logic with webpack. Hope that helps!

sergesemashko commented:
What's done:
- extracted logic into separate LintDirtyModulesOnlyPlugin, so it can keep its own state for runs count and previous timestamps
- installed testdouble.js
- removed webpack integration testing
- I hope, I've managed to create real "unit" tests :) (it was fun using testdouble)

@JaKXz, please, take a look once again

sergesemashko commented:
@JaKXz, sure, I'll take a look either today or tomorrow
